test/Base_Tests/src/Semantic/Multi_Value_Spec.enso

```diff
@@ -14,6 +35,61 @@ add_specs suite_builder =
             (c:Complex).re . should_equal 1.5
             (c:Float) . should_equal 1.5
 
+    suite_builder.group "Chain Multi Value" group_builder->
+        to_b_to_c obj =
+            b = obj : B
+            c = b : C
+            c
+
+        group_builder.specify "Multiple conversions of A work" <|
+            a = A
+            a_text = to_b_to_c a
+            a_text . to_text . should_equal "(C (B A))"
+
+        group_builder.specify "Multiple conversions of A & X work" <|
+            ax = A:A&X
+            a_text = to_b_to_c ax
+            a_text . to_text . should_equal "(C (B A & X))"
+
+        group_builder.specify "Multiple conversions of A & X work" <|
+            xa = A:X&A
+            a_text = to_b_to_c xa
+            a_text . to_text . should_equal "(C (B A & X))"
+
+        group_builder.specify "Multiple conversions of (A & X : A) work" <|
+            ax = A:A&X
+            a = ax : A
+            a_text = to_b_to_c a
+            a_text . to_text . should_equal "(C (B A & X))"
+
+        group_builder.specify "Multiple conversions of X fail" <|
+            x = X
+
+            # has to fail as X cannot be converted to B in to_b_to_c
+            Test.expect_panic Type_Error <|
+                to_b_to_c x
+
+        group_builder.specify "Multiple conversions of (A & X : X) work" <|
+            ax = A:A&X
+
+            ax.a . should_equal "a"
+            ax.x . should_equal "x"
+
+            x = ax : X
+
+            x.x . should_equal "x"
+            Test.expect_panic No_Such_Method <|
+                # method of A isn't visible for direct dispatch
+                x.a
+
+            # but the object can be converted to A and then
+            # method a is available
+            (x:A).a . should_equal "a"
+
+            # multivalue ax restricted to X cannot be converted to B in to_b_to_c
+            Test.expect_panic Type_Error <|
+                to_b_to_c x
```

Comment:
I'm most concerned about this case. First, I really appreciate this test - it really nicely illustrates the behaviour - that is great!

The way you implemented method dispatch makes sure that #9812 is implementable - **thank you**.

On the other hand, we can still have a value `x : X` and then we can cast it to `x : A` and the cast may succeed even though there is no (direct) relation between `X` and `A` (e.g. no subtyping relationship like `A <: X`). I know that this is how we discussed it - so I think this is correct. Moreover, I think that this is what we need to get the `Column & Integer_Column` facilities that we are aiming for future versions. So I think it is all good to go.

I just wanted to note that this will be a bit of a problem for future versions of static analysis. It works fine with method dispatch as noted above, but then one of the next steps, and IMO quite useful, was to add "lint" warnings about mismatched argument types - e.g.:
if I have a function `Bar.foo self (x : Integer) -> Integer` and I have code:
```
my_code (bar : Bar) (y : Text) =
    bar.foo y
```
One of the next steps would be that the inference knows that `bar.foo` is the call of the method `Bar.foo` and it could inspect the types of arguments. It will see that for the parameter `x : Integer` we have provided `y : Text`. In a world without MultiValueType, all I would do is check if there's a conversion `Integer.from (that : Text)` (in 3 scopes: Integer definition scope, Text definition scope and in the current scope of `my_code`); if there is no such conversion I could 'disprove' the conversion and report a warning. Which is in general useful because we can see that the check is guaranteed to fail. And this rids us of a whole class of errors where we e.g. used a wrong variable, or messed up the order of arguments (and now the types no longer match too). I assume we can all agree that this kind of "lint" is quite useful and many languages have something similar - obviously Java checks this statically because it has a simpler static type system, but also e.g. Python or TypeScript have an ability to issue similar warnings when we can see that the arguments are mismatched.

Now, with MultiValueType (not saying that this PR introduced this particular "issue") we have a problem. Even though my `y` is a `Text`, I don't know if under the hood it is not a multi-value that is actually `Text & Integer`. So even though in my current scope I don't have a `Integer.from (that : Text)` conversion, I can never be sure if some other scope does not have it, making it possible to construct a `Text & Integer` value that can then reach `y` introducing as `Text` but also being able to become `Integer` if asked to.

Now, I'm not saying this should be a reason to not proceed with the multi values. We definitely do want them because we plan to use them for useful functionalities in the libraries, and we will be using them sooner than type inference will be usable anyway. I'm however wondering if there's anything that we can do to somehow reconcile the two things - have the nice features of multi values but also be able to provide useful lints in the future.

## Possible Solutions

### Casts narrow-down the type _exactly_ but we can still pass mutli-value types around

We could make it so that a 'cast' `(x : T)` narrows down the type of `x` to be exactly `T`, meaning that if `x` was a multi value `T & U & V`, after going through `y = (x : T)`, then `y` is only `T` and a cast `y : U` will not succeed anymore (well, unless there is a conversion in scope - but the multi values are interesting where conversions _are not in scope_). So the cast `x:T` kind of 'cuts off' any additional multi values asserting the exact type.

This clearly makes the argument type checking linting possible again. But what about practical uses cases, what can we do to still be able to use multi values in presence of type ascriptions? Well I propose to mark multi-value types. For example, a most generic thing could be to mark that a type is `T` but may also contain other types through multi-value by marking it as `T & Any`. `(x : T & Any)` would essentially be what `x:T` means now - a cast that assures that `x` has type `T`, but it does not "cut off" any other types that a multi-value may contain. Thus, for our `Column` usage, our functions, like `Table.at` would be returning `Column & Any` meaning that they return  `Column` type with some possible 'mixed-in' types like `Column_Integer`.

I like this possible solution most, because it allows us to clearly state where we expect multi-value types and where the type is known to be exactly. In most cases, e.g. `Text` or `Integer` arguments, we won't really be expecting the multi-value behaviour, so (I think) we should be fine with the "cutting off" of types; and wherever we want the behaviour we can mark it explicitly in the return type. 

> Of course, `x : Any` is a "trivial cast" and it wouldn't do any "cutting off" as all values (also multi-) fit `Any`.

#### Narrowing with more precise types

Optionally, we could go even further and make the types a bit more precise. In methods like `Table.at` we actually know what the possible of mix-in values is, so we could make the return type more precise by noting that it returns `Column & (Integer_Column | Text_Column | ...)` meaning that we return a `Column` value that has one of the listed 'extra' types mixed-in. This gives even more information to the type inference that could be used in the future.

Of course such detailed and long types would be unwieldly to write and maintain in many function signatures. But we could fix that by introducing type aliases that would allow us to define `alias Column_Any = Column & (Integer_Column | Text_Column | ...)` and then use `Column_Any` in the return type. I think there was some idea of introducing type aliases some time ago, so this could prove useful.

### Limited lints by checking what multi-value types are at all possible.

If I understood correctly what @jdunkerley suggested when we were chatting about this. As noted above, with multi-value types even if we have `x : T` and we are expecting argument of type `U`, even if there is no conversion `T -> U` in scope, we cannot rule out that originally `x` was a multi-value of type `T & U` and thus it can be successfully casted. However, if we prove that in the whole codebase  (not even in current scope, but in all possible scopes) there is no conversion `T -> U`, that would mean that there is no way to 'organically' construct a value of type `T & U` and thus we can report that whenever we have a value `T` and expect `U`, that is a 'type error'.

This narrows down the set of type errors we can report, but it still allows us to report the error in many useful cases.

The problem with this approach is that it requires 'global' knowledge of available conversions - as to be able to report a type error, we need to be sure that no conversion exists between the two types, not only in current scope, but in no scope whatsoever. This means, that we cannot typecheck library Base in itself - we need to typecheck it in context of a whole running project, and importing more modules may mean that what used to be type error in Base, no longer is, because an external module may provide the 'missing' conversion. 

In such scenario the libraries are no longer self-contained. I don't think this will work well. If we wanted to make it work, perhaps we'd need to limit that a conversion for a given type can only be defined in the same library as one defining that type. Anyway, I feel like this approach would need some further design.

### Other solutions

If the solutions above are not satisfactory, perhaps we can think of something else? Let's discuss - what are the drawbacks of the solutions above and what else can we do to make it work better?

Reply:
Thank you Radek for such a detailed review and description of the...

> add "lint" warnings about mismatched argument types 

...problem. Yes, it deserves some solution. The first solution of yours - e.g. `Column & Any` was on my list of ideas in the past as well. The syntax is weird, but being able to explicitly say _"this is Column with some extensions"_ is the best _static analysis_ could hope for.

## More Solutions

#### Just a Warning

The _static analysis_ could just **emit a warning** - signaling a fishy code that is very likely broken, but could work under some circumstances (if one knows what one is doing). Given that _static analysis_ is meant as an optional _"linting"_ step, it may not be that bad solution.

#### Argument Cast vs. Explicit Cast

The automatic conversion/cast when passing an argument into just being invoked function doesn't have to be 100% equal to explicit (e.g. `x:Integer`) cast. We may decide that:
- automatic cast considers only types up to `methodDispatchTypes` - e.g. only selects from _"has been cast to"_ types and available conversions
- explicit cast considers all types - e.g. types the value _"can be cast to"_

This would allow _static analysis_ to reason about the types when passing a value as a function argument, but it would still give the IDE a chance to _cast multi value_ to one of its _extra types_.

Reply:
Short summary of our discussion with @JaroslavTulach and @jdunkerley 

We agreed that both solutions:
- Casts narrow-down the type exactly but we can still pass mutli-value types around
- Argument Cast vs. Explicit Cast
seem viable.

After some discussion we chose the latter. If I understood correctly, it's biggest advantage was the possibility of 'tunneling' a value through various function calls and later still being to access the additional types using an explicit cast. I.e. passing `x:A&B` through `f y:A` and then within `f` using explicit cast to see that it also has `B` in it. With the former solution, the `B` would be cut off and impossible to 'bring back', to avoid cutting it off we'd need to modify signature `f` to `f y:A&Any`. This tunneling as @JaroslavTulach stated gives us behaviour similar to open classes in Object Oriented languages - it allows the `A` class to be subclassed by `B` and passed through functions that expect `A`. Then the explicit cast acts like an `instanceof` check that allows to extract a 'more precise' type from the `A` instance.

I was unsure if subclassing analogy is exactly precise in Enso, because we can use conversions to create ad-hoc relationships: I can in my own module introduce conversion `Text.from Integer` that allows me to create instances `Integer&Text` where the 2 types were not related before. Real subclasses are related to each other 'by construction'. But nonetheless, it gives us a good analogy on how we plan to use it - we want to be able to "extend" some types, i.e. have a base `Column` type and then extend it with numeric functions by adding `& Integer_Column`. In this practical use-case the subclassing analogy does indeed prove useful.

----

This PR is already splitting our types into "2 parts": a visible one (used for method dispatch) and hidden one (possible to bring to be visible by the usage of explicit casts).

We want to change the behaviour so that argument-level checks (input arguments as well as return type) only act on the _visible_ part. This way we can continue doing static analysis, because the 'hidden' part of the type.

If we want to extract the hidden part (a more precise subclass in the analogy above), we need to do the explicit cast - similar to Java's `instanceof`. This cast can be done in 2 ways:
- either an explicit expression `a = b : T`
- or using `case of` - we want the case of statement to be able to extract the hidden part from multi values (I'm not sure if it was already capable of it or not - let's add some tests if we don't have).

The explicit cast is a "dynamic"/runtime cast - static analysis will not check if it is possible succeed or not, because we know it can always succeed depending on actual runtime values. But then static analysis will be able to rely on the result of the cast - if it succeeds the subsequent code can assert new type information about the casted value.

Reply:
At the beginning of today's discussion we selected only two viable approaches:
- Casts narrow-down the type exactly but we can still pass mutli-value types around
- Argument Cast vs. Explicit Cast

Both give us the needed _flexibility_ of multi values and reasonably defined _static checking_. The difference between those two is _tunneling_ or [teleportation](http://wiki.apidesign.org/wiki/Teleinterface). If there is a method:
```ruby
use_column c:Column callback:(Any -> Any) = callback c
```
then in the _casts narrow-down the type exactly_ approach the `callback` function receives `final class Column` (in the Java 1.4 OOP terminology) and there is no way to cast it to anything else. One would have to declare the `c:Column&Any` to allow also _subclasses_ of `Column` to go thru.

In the _Argument Cast vs. Explicit Cast_ this _openness to subclasses_ is automatic. When one writes `c:Column` one actually means `class Column` **and subclasses** (in the Java 1.4 terminology). E.g. this supports the [teleportation](http://wiki.apidesign.org/wiki/Teleinterface) by default. A special multi value _"subclass"_ can flow thru many `c:Column` checks without any issues just to be later tested to:
```ruby
case c of
  ic:Integer_Column -> ic.signum
  _ -> 0
```

At the end of the meeting we decided we'll go with the [OOP](http://wiki.apidesign.org/wiki/OOP)-like approach where `c:Column` is _open to subclasses_ ([Lišková would be happy](https://en.wikipedia.org/wiki/Liskov_substitution_principle)), but one has to do _explicit_ cast or `case of` to extract the _extra types_ associated with a value.

Reply:
> In pseudocode, I'd write the algorithms as:

Instead of reviewing the real [AbstractTypeCheckNode and subclasses](https://github.com/enso-org/enso/pull/11600/files#diff-9a0d1cc521db14e10790ebfeea3c54df20e413ace4c75a57182434b17b2ebd0aL16) we turn this review into a review of pseudocode. Ach. Pseudocode cannot be executed, debugged and has no tests... unlike this PR.

> @JaroslavTulach do you agree with the pseudocode above? 

Probably. But I am neither a computer nor a testing framework to interpret it correctly.

> What do you think about the variants for explicit cast? I lean towards Variant 1 as it is more explicit

Yes, I believe we are implementing Variant 1.

The drawback of Variant 2 is _explosion of intersection types_. Variant 1 uses _only intersection types that explicitly spelled_ in the source code somewhere. The Variant 2 can create _intersection types power set_  by programmatically adding yet another type to a value. An efficient interpreter needs the types of values to stabilize. When I imagine where the _extending of types of a value_ can lead, I stopped considering Variant 2.

I find Variant 1 more conservative, more _"interpreter friendly"_ and still usable enough.

Reply:
I've just created [intersection types docs](https://github.com/enso-org/enso/blob/wip/jtulach/ComplexAndFloat11482/docs/types/intersection-types.md). Let's make sure the summary of this discussion is available [there](https://github.com/enso-org/enso/blob/wip/jtulach/ComplexAndFloat11482/docs/types/intersection-types.md). Feel free to [edit the text](https://github.com/enso-org/enso/blob/wip/jtulach/ComplexAndFloat11482/docs/types/intersection-types.md), but please let us avoid pasting pseudocode there ;-)